Add redirect support for iOS and Android.

jeremyfaller opened the pull request:
Reviewers:

Please take a close look at caching and the terraform config work here. For caching, I went with purely the DB cache -- not going any deeper. If we need more, please LMK.


Fixes #580

<!-- Please include the 'why' behind your changes if no issue exists -->
## Proposed Changes

* Creates a pair of endpoints for iOS and Android to get app redirect information
* Adds the a few new fields to `authorized_apps` to hold data needed for the redirection service.

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
Redirection service now publishes the iOS and Android redirection metadata.
APIKeys now includes support for specifying the needed iOS and Android metadata for redirection.
```

sethvargo reviewed (changes requested):
I'm concerned about tying the API key creation to a particular app/os architecture. This might be better as its own table. We should also be imploring more aggressive caching of these JSON responses, since every device will request them automatically like a favicon.

jeremyfaller commented on cmd/server/assets/apikeys/new.html:
```diff
@@ -49,13 +49,67 @@ <h1>New API key</h1>
             {{end}}
           </div>
 
+          <div class="form-group">
```
As written, yes. I can update to handle both.

jeremyfaller commented on pkg/database/migrations.go:
```diff
@@ -32,6 +32,22 @@ func (db *Database) getMigrations(ctx context.Context) *gormigrate.Gormigrate {
 	options := gormigrate.DefaultOptions
 
+	// Helper function to safely add columns to the database.
+	addColumn := func(tx *gorm.DB, table, name, colType, defaultVal string) error {
```
I put it there because it seems a common error-prone motif. I'll move it.

icco commented on pkg/controller/associated/ios.go:
```diff
@@ -0,0 +1,84 @@
+
+import "github.com/google/exposure-notifications-verification-server/pkg/database"
+
+type IosData struct {
```
nit: I would expect this to be IOSData given how you're capitalizing it in other places.

icco commented on pkg/controller/apikey/create.go:
```diff
@@ -16,19 +16,62 @@ package apikey
+	Type  database.APIUserType `form:"type"`
+	OS    database.OsType      `form:"os"`
+	AppID string               `form:"appId"`
+	Sha   string               `form:"sha"`
```
nit: I'd expect this to be SHA

icco commented on terraform/service_redirect.tf:
```diff
@@ -31,6 +31,19 @@ resource "google_service_account_iam_member" "cloudbuild-deploy-enx-redirect" {
+
 resource "google_project_iam_member" "enx-redirect-observability" {
   for_each = toset([
     "roles/cloudtrace.agent",
```
Github isn't letting me leave a comment, but in the `dynamic "env" {` block you need to add the configs for the cache and database.

jeremyfaller commented on cmd/server/assets/apikeys/show.html:
```diff
@@ -57,6 +57,32 @@ <h1>{{$authApp.Name}} API key</h1>
           {{end}}
         </div>
+
+        <strong>OS</strong>
```
done

jeremyfaller commented:
Seth:

I think I addressed your comments here by improving the editing of the keys (to allow them to move between OS's etc.).

I still need to improve caching (of the JSON responses), so this isn't quite complete. But I will either address that in a followon, or this CL can go. Up to you.

jeremyfaller commented:
Caching is in now as well.

Unless you'd like to see something different from the design, @sethvargo, this is ready for your review.

jeremyfaller commented:
unintentionally closed.

mikehelmick commented on cmd/server/assets/apikeys/new.html:
```diff
@@ -49,6 +49,8 @@ <h1>New API key</h1>
             {{end}}
           </div>
 
+		  {{template "os.html" .}}
```
I'm not sure this belongs in the API keys page.

Seems like it should be a separate realm level setting

jeremyfaller replied:
Ok, so it looks like the original proposal I wrote in the issue (to put this in Health Authority) won't work. I'll move to a separate table.

/hold

mikehelmick replied:
My fault - I didn't get to the proposal on the issue before the PR was sent

jeremyfaller replied:
Moved separately.

mikehelmick commented on pkg/database/authorized_app.go:
```diff
@@ -61,8 +70,22 @@ type AuthorizedApp struct {
 	APIKeyType APIUserType `gorm:"default:0"`
+
+	// OS is the type of the application we're using (eg, iOS, Android).
+	OS OsType `gorm:"os,type:int"`
```
Having these on API keys isn't going to work - because API keys will be rotated eventually.
the current positioning would actually let a realm have multiple apps per-platform claimed, and I don't think we want that either.

This information should only be set once on the realm.

jeremyfaller replied:
Moved to a non-per-realm basis.

jeremyfaller commented:
/hold

peetsasaki commented on cmd/server/assets/apikeys/os.html:
```diff
@@ -0,0 +1,76 @@
+
+  <div class="form-label-group" id="androidDIV">
+    <input type="text" id="androidAppID" name="androidAppID" class="form-control{{if $authApp.ErrorsFor "androidAppID"}} is-invalid{{end}}" value="{{$authApp.AndroidAppID}}" placeholder="Application ID" autofocus>
+    <label for="androidAppID">Android Application ID</label>
```
nit: By "Application ID" do we mean package name, e.g. "com.example.app"? If so, might call it "Package Name" in the UI since that's understood by Android devs and consistent with Google's docs re app links https://developer.android.com/training/app-links/verify-site-associations

jeremyfaller replied:
I've moved to BundleID, LMK if something else is preferrable.

peetsasaki commented on cmd/server/assets/apikeys/os.html:
```diff
@@ -0,0 +1,76 @@
+      {{joinStrings ($authApp.ErrorsFor "sha") ", "}}
+    </div>
+    {{end}}
+  </div>
```
I can see in other files that the underlying storage & code support multiple SHA hashes per app, which is great. I'm not familiar with this html templating engine, so I can't tell if the UI intends to support multiple hashes "top to bottom". If so great, but if not, it might be good to do so, to help smooth integration with the app, given developers will need to test with a debug-keyed build prior to publishing a release-keyed build.

jeremyfaller replied:
Do we require multiple SHA hashes per app? Currently, I don't support this, but could.

jeremyfaller replied:
Moved to supporting multiple SHAs for an android app.

peetsasaki commented on cmd/server/assets/apikeys/os.html:
```diff
@@ -0,0 +1,76 @@
+
+  <div class="form-label-group" id="iosDIV">
+    <input type="text" id="iosAppID" name="iosAppID" class="form-control{{if $authApp.ErrorsFor "iosAppID"}} is-invalid{{end}}" value="{{$authApp.IOSAppID}}" placeholder="Application ID" autofocus>
+    <label for="iosAppID">iOS Application ID</label>
```
nit: Similarly to below: if "Application ID" here means "bundle identifier", it might be good to call it that in UI.

jeremyfaller replied:
I'll update the UI to move to bundle id for android.

jeremyfaller commented:
There is one outstanding issue before this is ready for review -- I need clarification on whether or not an android app could have multiple bundle ids.

@peetsasaki please comment

jeremyfaller commented:
More research leads me to believe android does need multiple SHAs per app. Updating the CL. continue to hold.

jeremyfaller commented:
/unhold

This is ready for review.

sethvargo commented on cmd/server/assets/mobileapp/edit.html:
```diff
@@ -0,0 +1,57 @@
+          <div class="form-label-group">
+            <input type="text" id="name" name="name" class="form-control{{if $app.ErrorsFor "name"}} is-invalid{{end}}" value="{{$app.Name}}" placeholder="Application name" autofocus>
+            <label for="name">Application name</label>
+            {{if $app.ErrorsFor "name"}}
```
errorable

sethvargo commented on cmd/server/assets/mobileapp/app.html:
```diff
@@ -0,0 +1,52 @@
+{{$app := .app}}
```
And it should probably be `_form.html`  for consistency since this appears to be a partial.

jeremyfaller replied:
The `{{template "app.html" .}}` worked as it was, but NBD your way, and more consistent!

sethvargo commented on cmd/server/assets/mobileapp/index.html:
```diff
@@ -0,0 +1,84 @@
+  <main role="main" class="container">
+    {{template "flash" .}}
+
+    <h1>Mobile Apps</h1>
```
```suggestion
    <h1>Mobile apps</h1>
```

sethvargo commented on cmd/server/assets/mobileapp/edit.html:
```diff
@@ -0,0 +1,57 @@
+  <main role="main" class="container">
+    {{template "flash" .}}
+
+    <h1>Edit Mobile App</h1>
```
```suggestion
    <h1>Edit mobile app</h1>
```

sethvargo commented on cmd/server/assets/mobileapp/app.html:
```diff
@@ -0,0 +1,52 @@
+  </div>
+
+  <script>
+    function osSelect() {
```
We should bind this to the select box via jquery or javascript instead of relying on the HTML author to add `osSelect` as an onChange handler.

sethvargo commented on cmd/server/assets/mobileapp/edit.html:
```diff
@@ -0,0 +1,57 @@
+
+		  {{template "app.html" .}}
+
+          <button type="submit" class="btn btn-primary btn-block">Update Mobile App</button>
```
```suggestion
          <button type="submit" class="btn btn-primary btn-block">Update mobile app</button>
```

sethvargo commented on cmd/server/assets/mobileapp/new.html:
```diff
@@ -0,0 +1,50 @@
+  <main role="main" class="container">
+    {{template "flash" .}}
+
+    <h1>New Mobile App</h1>
```
```suggestion
    <h1>New mobile app</h1>
```

sethvargo commented on cmd/server/assets/mobileapp/app.html:
```diff
@@ -0,0 +1,52 @@
+{{$app := .app}}
```
These assets should be under mobileapp**s** `mobileapps`

sethvargo replied:
And the templates should be name as-such.

sethvargo commented on cmd/server/assets/mobileapp/new.html:
```diff
@@ -0,0 +1,50 @@
+          <div class="form-label-group">
+            <input type="text" id="name" name="name" class="form-control{{if $app.ErrorsFor "name"}} is-invalid{{end}}" value="{{$app.Name}}" placeholder="Application name" autofocus>
+            <label for="name">Application name</label>
+            {{if $app.ErrorsFor "name"}}
```
errorable

sethvargo commented on cmd/server/assets/mobileapp/new.html:
```diff
@@ -0,0 +1,50 @@
+    <div class="card mb-3 shadow-sm">
+      <div class="card-header">Details</div>
+      <div class="card-body">
+        <form method="POST" action="/mobileapp">
```
You need to add `class="floating-form"` to get the good HTML floating forms

sethvargo commented on cmd/server/assets/mobileapp/show.html:
```diff
@@ -0,0 +1,76 @@
+      <a href="/mobileapp/{{$app.ID}}/edit">Edit</a>
+    </p>
+    <p>
+      Here is information about the Mobile App.
```
```suggestion
      Here is information about the mobile app.
```

sethvargo commented on cmd/server/assets/mobileapp/show.html:
```diff
@@ -0,0 +1,76 @@
+          {{end}}
+        </div>
+
+		{{if (eq $app.OS .iOS)}}
```
TvsS

sethvargo commented on cmd/server/assets/mobileapp/show.html:
```diff
@@ -0,0 +1,76 @@
+          </div>
+        {{end}}
+
+		{{if (eq $app.OS .android)}}
```
TvsS

sethvargo commented on cmd/server/main.go:
```diff
@@ -301,6 +302,29 @@ func realMain(ctx context.Context) error {
 
+	// mobileapp
+	{
+		sub := r.PathPrefix("/mobileapp").Subrouter()
```
This should be `/mobileapps` or, better `/mobile-apps`

sethvargo commented on cmd/server/main.go:
```diff
@@ -301,6 +302,29 @@ func realMain(ctx context.Context) error {
+		sub.Handle("", mobileController.HandleIndex()).Methods("GET")
+		sub.Handle("", mobileController.HandleCreate()).Methods("POST")
+		sub.Handle("/new", mobileController.HandleCreate()).Methods("GET")
+		sub.Handle("/{id}/edit", mobileController.HandleUpdate()).Methods("GET")
```
Add constraints on these IDs (we haven't backported to the other things yet), but they should be:

```
{id:[0-9]+}
```

sethvargo commented on pkg/controller/associated/android.go:
```diff
@@ -0,0 +1,54 @@
+func (c *Controller) getAndroidData() ([]AndroidData, error) {
+	apps, err := c.db.FindActiveAppsByOS(database.OSTypeAndroid)
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, fmt.Errorf("failed to get android data: %w", err)
```

sethvargo commented on pkg/controller/associated/android.go:
```diff
@@ -0,0 +1,54 @@
+			Target: Target{
+				Namespace:    "android_app",
+				PackageName:  apps[i].AppID,
+				Fingerprints: strings.Split(apps[i].SHA, "\n"),
```
Yea, I think we need some more intelligence and sanitization here. We should be a helper and be tolerant to blank fingerprints, CSV instead of NLSV, etc.

sethvargo commented on pkg/controller/associated/index.go:
```diff
@@ -0,0 +1,116 @@
+	h      *render.Renderer
+}
+
+var ttl = 30 * time.Minute
```
const?

sethvargo replied:
and doc

sethvargo commented on pkg/controller/associated/index.go:
```diff
@@ -0,0 +1,116 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		key := r.URL.String()
+
+		// Check the cache.
```
You should be able to use `c.cacher.Fetch` here instead of a manual read, create, write. 

```go
var data IOSData
if err := c.cacher.Fetch(ctx, key, &data, ttl, func() (interface{}, error){
  return c.getIosData()
}); err != nil {
  // handle error
}
```

sethvargo commented on pkg/controller/associated/index.go:
```diff
@@ -0,0 +1,116 @@
+)
+
+type Controller struct {
+	ctxt   context.Context
```
Don't store contexts on structs - it's a [Go no-no](https://golang.org/pkg/context/):

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:

sethvargo commented on pkg/controller/associated/index.go:
```diff
@@ -0,0 +1,116 @@
+		// Check the cache.
+		{
+			var data IOSData
+			if err := c.cacher.Read(c.ctxt, key, &data); err == nil {
```
This context should be pulled from the request (`r.Context()`) or passed in to `HandleIos`, depending on who we want to control cancelation.

sethvargo commented on pkg/controller/associated/index.go:
```diff
@@ -0,0 +1,116 @@
+		}
+
+		// Cache the data.
+		if err := c.cacher.Fetch(c.ctxt, key, &data, ttl, func() (interface{}, error) {
```
Yea, I think you can simplify this greatly but just calling getIosData() in here an removing everything above.

sethvargo commented on pkg/controller/associated/index.go:
```diff
@@ -0,0 +1,116 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		key := r.URL.String()
+
+		// Check the cache
```
Same notes about cache and context.

sethvargo commented on pkg/controller/mobileapp/create.go:
```diff
@@ -0,0 +1,171 @@
+
+// cleanSHA cleans an input string.
+func cleanSHA(input string) string {
+	input = strings.ToUpper(input)
```
This is pretty expensive... it probably doesn't matter for this use case, but ReplaceAll is an O(n) op and we're calling it 5x

sethvargo replied:
Also, it feels like `clean` and `verify` should be the same function.

sethvargo commented on pkg/controller/mobileapp/create.go:
```diff
@@ -0,0 +1,171 @@
+		}
+
+		// Verify the form.
+		if err := form.verify(); err != nil {
```
Hmm - what does this do?

sethvargo commented on pkg/controller/mobileapp/create.go:
```diff
@@ -0,0 +1,171 @@
+
+		// Store the app name in the session so we can properly display it on
+		// the next page.
+		session.Values["appName"] = form.Name
```
Hmm - why do we need this on the next page? It's generally frowned upon to store things in the session that aren't long-term persisted. We use flash for short-term things.

sethvargo commented on pkg/controller/mobileapp/create.go:
```diff
@@ -0,0 +1,171 @@
+}
+
+// renderNew renders the new page.
+func (c *Controller) renderNew(ctxt context.Context, w http.ResponseWriter, app *database.MobileApp) {
```
context should be named `ctx` throughout (that's a Go-ism)

sethvargo commented on pkg/controller/mobileapp/disable.go:
```diff
@@ -0,0 +1,62 @@
+
+		if err := app.Disable(c.db); err != nil {
+			flash.Error("Failed to disable mobile app: %v", err)
+			http.Redirect(w, r, "/mobileapp", http.StatusSeeOther)
```
Need a `return` after the redirect.

sethvargo commented on pkg/controller/mobileapp/show.go:
```diff
@@ -0,0 +1,78 @@
+		app, err := realm.FindMobileApp(c.db, vars["id"])
+		if err != nil {
+			if database.IsNotFound(err) {
+				logger.Debugw("mobile app does not exist", "id", vars["id"])
```
Drop this log message

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+	RealmID uint `gorm:"unique_index:realm_app_name"`
+
+	// OS is the type of the application we're using (eg, iOS, Android).
+	OS OSType `gorm:"os,type:int"`
```
```suggestion
	OS OSType `gorm:"column:os; type:int"`
```

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+	Errorable
+
+	// Name is the name of the app.
+	Name string `gorm:"name,varchar(512);unique_index:realm_app_name"`
```
```suggestion
	Name string `gorm:"column:name; type:citext; unique_index:realm_app_name;"`
```

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+	OS OSType `gorm:"os,type:int"`
+
+	// IOSAppID is a unique string representing the app.
+	AppID string `gorm:"app_id,type:varchar(512)"`
```
```suggestion
	AppID string `gorm:"column:app_id; type:varchar(512);"`
```

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+	// SHA is a unique hash of the app.
+	// It is only present for Android devices, and should be of the form:
+	//   AA:BB:CC:DD...
+	SHA string `gorm:"sha,type:text"`
```
```suggestion
	SHA string `gorm:"column:sha; type:varchar(TODO);"`
```

sethvargo replied:
We know the max length of the SHA, right?

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+	if err := db.db.
+		Model(&MobileApp{}).
+		Where("os = ?", os).
+		Where("deleted_at IS NULL").
```
This is part of the default query - you can drop it.

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// CreateMobileApp adds a mobile app to the DB.
+func (r *Realm) CreateMobileApp(db *Database, app *MobileApp) error {
```
I'm not sure we want mutating functions to live off other database models. This should probably live off of `Database`.

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+
+// SaveMobileApp saves the authorized app.
+func (db *Database) SaveMobileApp(r *MobileApp) error {
+	if r.Model.ID == 0 {
```
Don't need this - just call `db.db.Save(r).Error`

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+
+// Disable disables the mobile app.
+func (a *MobileApp) Disable(db *Database) error {
+	if err := db.db.Delete(a).Error; err != nil {
```
Just return `db.db.Delete(a).Error`

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+
+// Enable enables the mobile app.
+func (a *MobileApp) Enable(db *Database) error {
+	if err := db.db.Unscoped().Model(a).Update("deleted_at", nil).Error; err != nil {
```
Just return

sethvargo commented on pkg/database/mobile_app.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// FindActiveAppsByOS finds all authorized by their OS.
+func (db *Database) FindActiveAppsByOS(os OSType) ([]*MobileApp, error) {
```
This should be `List`, not `Find`

sethvargo commented on pkg/database/realm.go:
```diff
@@ -571,6 +571,38 @@ func (r *Realm) FindAuthorizedApp(db *Database, id interface{}) (*AuthorizedApp,
+func (r *Realm) ListMobileApps(db *Database) ([]*MobileApp, error) {
+	var apps []*MobileApp
+	if err := db.db.
+		Unscoped().
```
Why unscoped? that would include deleted.

jeremyfaller replied:
I went with the same approach of being able to resurrect apps (like you can resurrect keys). So, this is A-OK (the UI supports it, etc.)

![image](https://user-images.githubusercontent.com/3063869/94289564-e25ff100-ff26-11ea-8983-adecf7fdd1cc.png)

sethvargo commented on pkg/database/realm.go:
```diff
@@ -571,6 +571,38 @@ func (r *Realm) FindAuthorizedApp(db *Database, id interface{}) (*AuthorizedApp,
+		Related(&apps).
+		Error; err != nil {
+		if IsNotFound(err) {
+			return nil, nil
```
```suggestion
			return apps, nil
```

We want to give people the empty list instead of nil.

jeremyfaller replied:
hehe. they're _mostly_ the same, but you're right -- not exactly.

sethvargo commented on pkg/database/realm.go:
```diff
@@ -571,6 +571,38 @@ func (r *Realm) FindAuthorizedApp(db *Database, id interface{}) (*AuthorizedApp,
+	if err := db.db.
+		Unscoped().
+		Model(MobileApp{}).
+		Order("LOWER(name) ASC").
```
You're only expecting one record - no need to order

sethvargo replied:
And there's a unique index

sethvargo commented on terraform/service_redirect.tf:
```diff
@@ -69,6 +137,8 @@ resource "google_cloud_run_service" "enx-redirect" {
             local.gcp_config,
             local.enx_redirect_config,
+            local.cache_config,
+            local.database_config,
```
I think you nee csrf and some other things too?

jeremyfaller replied:
@icco Can you comment here? I don't want to mess this up, like I did jwks, and I am **NOT** the person you want to leave as a babysitter on this one. :)

icco replied:
Looking at cmd/enx-redirect/main.go I don't think you should need any other configs.

jeremyfaller commented on cmd/enx-redirect/main.go:
```diff
@@ -58,14 +61,14 @@ func main() {
 func realMain(ctx context.Context) error {
 	logger := logging.FromContext(ctx)
 
-	config, err := config.NewRedirectConfig(ctx)
```
Sure, I'll followup with all the other servers. I think I can cobble a `gofmt` script to do it.

jeremyfaller replied:
https://github.com/google/exposure-notifications-verification-server/issues/674

I'll file similar for ENS.

sethvargo commented:
/lgtm
